datafusion/optimizer/src/simplify_expressions.rs

```diff
@@ -790,6 +791,43 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                 right,
             } if !info.nullable(&left)? && left == right => lit(1),
 
+            //
+            // Rules for Modulo
+            //
+
+            // A % null --> null
+            BinaryExpr {
+                left: _,
+                op: Modulo,
+                right,
+            } if is_null(&right) => *right,
+            // null % A --> null
+            BinaryExpr {
+                left,
+                op: Modulo,
+                right: _,
+            } if is_null(&left) => *left,
+            // A % 1 --> 0
+            BinaryExpr {
+                left,
+                op: Modulo,
+                right,
+            } if !info.nullable(&left)? && is_one(&right) => lit(0),
+            // A % A --> 0
+            BinaryExpr {
+                left,
+                op: Modulo,
+                right,
+            } if !info.nullable(&left)? && left == right => lit(0),
```

Comment:
I prefer removing `A % A` because `A` might contain `0` and `0 % 0 = 0` is wrong

Reply:
Rather than removing, I think I can move the DivideByZero condition above `A % A` so that code does not evaluate `A % A` if A is zero. I'll make that change and update this PR. Let me know if you see any issues.

Reply:
Moving `A % 0` above `A % A` is reasonable. 👍

However, we should also consider `A` is not a literal `0` but a column that contains `0`. For example:
```sql
> create table t as select 0 as a;

> select a % a from t;
DivideByZeroError
```
In this context, we cannot simplify `a % a` to zero.

Reply:
Removed `A % A`